### PR TITLE
Add lower electron temperature cut-off to ThomasFermi model

### PR DESCRIPTION
**Pull request by n01r**
Depending on the material the Thomas-Fermi prediction for the average charge state can be unphysically high. For some materials it predicts non-zero charge states at 0 temperature. That can be a reasonable approximation for metals and their electrons in the conduction band. In the general case there should be no free electrons at 0 temperature, though.
    
Therefore this commit implements a lower temperature cutoff.
The `detailedBalanceThomasFermi` function now accepts the parameters

- `temperature` in eV
- `massDensity` in g/cm^3

just as the fitting model described by More (1985).

See the Thomas-Fermi prediction of the charge state of the constituents of polystyrene, carbon and hydrogen.
tf_prediction_plastic_t 0

- [x] run a simple test: *electron temperature cutoff at 1 eV - material: copper at solid density of 8.9 g/cm³*   
    - once run with 0 eV electron temperature - TF prediction: <Z> = 4.40
    - one run with 5 eV: TF prediction for - TF prediction: <Z> = 4.66
    - **result** success: due to cutoff at 1 eV, no ionization in the 0 eV case while in the 5 eV case the average charge state amounts to <Z> = 4.66

## To Do

- [x] please rebase against dev
- [x] add the "ion density" cutoff set to "potential free electron density" comment in more detail in example
- [x] add plots and parts of the great PR descriptions to the model page

**update 2017-11-23**

This plot will automatically be created by sphinx.
tf_prediction_plastic_t 0_10_and_100ev

### CODE REVIEW

**Comment by ax3l** [comment]
thx, great work both PRs! ✨ 

@n01r offline discussion to do:
- [ ] please rebase against dev
- [ ] add the "ion density" cutoff set to "potential free electron density" comment in more detail in example
- [ ] add plots and parts of the great PR descriptions to the model page

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp [hunk 1]
```diff
@@ -166,13 +146,49 @@ namespace ionization
             uint32_t numNewFreeMacroElectrons = 0u;
 
             float_64 const densityUnit = static_cast< float_64 >( particleToGrid::derivedAttributes::Density( ).getUnit( )[ 0 ] );
+            float_64 const kinEnergyDensityUnit = static_cast<float_64>(particleToGrid::derivedAttributes::EnergyDensity().getUnit()[0]);
```
Comment (ax3l) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp]: if you are on it, you can add spaces in `<>` and `()` backets :)

File: docs/source/models/collisional_ionization.rst [hunk 2]
```diff
@@ -41,6 +52,16 @@ We therefore implemented additional conditions to mediate unphysical behavior bu
     It is strongly suggested to do approximations for **every** setup or material first.
     To that end, a parameter scan with [FLYCHK]_ can help in choosing a reasonable value.
 
+3. Low electron-temperature cutoff
+
+    Depending on the material the Thomas-Fermi prediction for the average charge
```
Comment (ax3l) [inline docs/source/models/collisional_ionization.rst]: In sphinx as in latex, please write one sentence per line independent of the line length.

File: docs/source/models/collisional_ionization.rst [hunk 3]
```diff
@@ -41,6 +52,16 @@ We therefore implemented additional conditions to mediate unphysical behavior bu
...
+    state can be unphysically high. For some materials it predicts non-zero
+    charge states at 0 temperature. That can be a reasonable approximation
+    for metals and their electrons in the conduction band.
+    Yet this cannot be generalized for all materials and therefore a cutoff should be explicitly defined.
+
+    - define via ``CUTOFF_LOW_TEMPERATURE_EV`` in ``ionizer.param``
```
Comment (ax3l) [inline docs/source/models/collisional_ionization.rst]: replace ionizer.param with
```
:ref:`ionizer.param <usage-params-extensions>`
```

File: docs/source/models/collisional_ionization_thomas-fermi_cutoffs.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        self.b_0 = -1.7630
+        self.b_1 = 1.43175
+        self.b_2 = 0.31546
+
+        self.c_1 = -0.366667
+        self.c_2 = 0.983333
+
+        self.alpha = 14.3139
+        self.beta = 0.6624
+        # print("initialized")
```
Comment (ax3l) [inline docs/source/models/collisional_ionization_thomas-fermi_cutoffs.py]: pls remove

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp [hunk 5]
```diff
@@ -63,40 +63,27 @@ namespace ionization
...
-            float_64 const kinEnergyUnit = kinEnergyDensityUnit / densityUnit;
-            float_64 const kinEnergy = kinEnergyDensity / density * kinEnergyUnit;
-            /** convert kinetic energy in J to "temperature" in eV by assuming an ideal electron gas
-             * E_kin = 3/2 k*T
-             */
-            constexpr float_64 convKinEnergyToTemperature = UNITCONV_Joule_to_keV * float_64(1.e3) * float_64(2./3.);
-            float_64 const temperature = kinEnergy * convKinEnergyToTemperature;
-
-            float_64 const T_0 = temperature/math::pow(protonNumber,float_64(4./3.));
+            float_64 const T_0 = temperature/math::pow( protonNumber,float_64( 4./3. ) );
```
Comment (ax3l) [inline include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp]: just when you are on it@ missing space after `,` and around operator `/`

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 6]
```diff
@@ -447,8 +447,25 @@ namespace thomasFermi
      * @note This cutoff value should be set in accordance to FLYCHK calculations,
      *       for instance! It is not a universal value and requires some preliminary
      *       approximations!
+     *
+     * unit: 1 / m^3
+     *
+     * default: ion number density = critical electron number density for lambda_laser = 800nm
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: 1) title: lower +ion+ density cutoff

"default" is a bit misleading here.
Maybe write:
```
example: 1.7422e27 as a hydrogen ion number density equal to the corresponding critical electron number density for a 800nm laser
```
```

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 7]
```diff
@@ -447,8 +447,25 @@ namespace thomasFermi
...
+     *
+     * The choice of the default is motivated by by the following:
+     * In laser-driven plasmas all dynamics in density regions below the
+     * critical electron density will be laser-dominated. Once ions of that density
+     * are ionized once the laser will not penetrate fully anymore and the as electrons are heated
+     * the dynamics will be collision-dominated.
+     */
+    constexpr float_X CUTOFF_LOW_DENSITY = 1.7422e27;
+
+    /** low temperature cutoff
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: low +electron+ temperature cutoff

File: include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp (hunk 5 above)
Comment (n01r) [reply include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp]: one day I'll get them all...

Comment (ax3l) [reply include/picongpu/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp]: hopefully, `clang-format` will get them all for us :) (with llvm 5.0+)

File: include/picongpu/simulation_defines/param/ionizer.param [hunk 8]
```diff
@@ -447,8 +447,25 @@ namespace thomasFermi
      * @note This cutoff value should be set in accordance to FLYCHK calculations,
      *       for instance! It is not a universal value and requires some preliminary
      *       approximations!
+     *
+     * unit: 1 / m^3
+     *
+     * example: 1.7422e27 as a hydrogen ion number density equal to the corresponding critical electron number density for an 800nm laser
```
Comment (ax3l) [inline include/picongpu/simulation_defines/param/ionizer.param]: note: such a long line ;)

**Pull request merged**